sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkGetFunctionsOperation.scala

```diff
@@ -0,0 +1,118 @@
+
+import org.apache.hadoop.hive.ql.security.authorization.plugin.{HiveOperationType, HivePrivilegeObjectUtils}
+import org.apache.hive.service.cli._
+import org.apache.hive.service.cli.operation.GetFunctionsOperation
+import org.apache.hive.service.cli.session.HiveSession
+import org.apache.thrift.TException
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql.SQLContext
+import org.apache.spark.util.{Utils => SparkUtils}
+
+/**
+ * Spark's own GetTablesOperation
+ *
+ * @param sqlContext SQLContext to use
+ * @param parentSession a HiveSession from SessionManager
+ * @param catalogName catalog name. null if not applicable
+ * @param schemaName database name, null or a concrete database name
+ * @param functionName function name pattern
+ */
+private[hive] class SparkGetFunctionsOperation(
+    sqlContext: SQLContext,
+    parentSession: HiveSession,
+    catalogName: String,
+    schemaName: String,
+    functionName: String)
+  extends GetFunctionsOperation(parentSession, catalogName, schemaName, functionName)
+    with Logging {
+
+  override def runInternal(): Unit = {
+    val statementId = UUID.randomUUID().toString
+    // Do not change cmdStr. It's used for Hive auditing and authorization.
+    val cmdStr = s"catalog : $catalogName, schemaPattern : $schemaName"
+    val logMsg = s"Listing functions '$cmdStr'"
+    logInfo(s"$logMsg with $statementId")
+    setState(OperationState.RUNNING)
+    // Always use the latest class loader provided by executionHive's state.
+    val executionHiveClassLoader = sqlContext.sharedState.jarClassLoader
+    Thread.currentThread().setContextClassLoader(executionHiveClassLoader)
+
+    val catalog = sqlContext.sessionState.catalog
+
+    if (isAuthV2Enabled) {
+      // get databases for schema pattern
+      val schemaPattern = convertSchemaPattern(schemaName)
+      var matchingDbs: Seq[String] = null
+      try {
+        matchingDbs = catalog.listDatabases(schemaPattern)
+      } catch {
+        case e: TException =>
+          setState(OperationState.ERROR)
+          HiveThriftServer2.listener.onStatementError(
+            statementId, e.getMessage, SparkUtils.exceptionString(e))
+          throw new HiveSQLException(e)
+      }
+      // authorize this call on the schema objects
+      val privObjs =
+        HivePrivilegeObjectUtils.getHivePrivDbObjects(seqAsJavaListConverter(matchingDbs).asJava)
+      authorizeMetaGets(HiveOperationType.GET_FUNCTIONS, privObjs, cmdStr)
+    }
+
+    HiveThriftServer2.listener.onStatementStart(
+      statementId,
+      parentSession.getSessionHandle.getSessionId.toString,
+      logMsg,
+      statementId,
+      parentSession.getUsername)
+
+    try {
+      val functionPattern = CLIServiceUtils.patternToRegex(functionName)
+      if ((null == catalogName || "".equals(catalogName))
+        && (null == schemaName || "".equals(schemaName))) {
+        catalog.listFunctions(catalog.getCurrentDatabase, functionPattern).foreach {
```

Comment:
Actually, I'm confused about this code. Maybe it should be:
```scala

val functionPattern = CLIServiceUtils.patternToRegex(functionName)
matchingDbs.foreach { schema =>
  catalog.listFunctions(catalog.getCurrentDatabase, functionPattern).foreach {
    case (functionIdentifier, _) =>
      val rowData = Array[AnyRef](
        DEFAULT_HIVE_CATALOG, // FUNCTION_CAT
        schema, // FUNCTION_SCHEM
        functionIdentifier.funcName, // FUNCTION_NAME
        "", // REMARKS
        DatabaseMetaData.functionResultUnknown.asInstanceOf[AnyRef], // FUNCTION_TYPE
        "") // SPECIFIC_NAME
      rowSet.addRow(rowData);
}
```
But it's the logic of Hive: https://github.com/apache/hive/blob/rel/release-3.1.1/service/src/java/org/apache/hive/service/cli/operation/GetFunctionsOperation.java#L101-L119

Reply:
I think we should cover the case of functions that don't have a schema (null) which is basically what Hive's implementation seems to do, as well as the functions associated with a given schema which is what your code snippet above seems to do. Could you combine both?

https://docs.microsoft.com/en-us/sql/connect/jdbc/reference/getfunctions-method-sqlserverdatabasemetadata?view=sql-server-2017

Reply:
May be we do not need to care about `catalog`:
https://github.com/pgjdbc/pgjdbc/blob/17c4bcfb59e846c593093752f2e30dd97bb4b338/pgjdbc/src/main/java/org/postgresql/jdbc/PgDatabaseMetaData.java#L2612-L2649